src/Type/Php/ArraySearchFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,90 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\Constant\ConstantArrayType;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\ConstantScalarType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\NullType;
+use PHPStan\Type\ObjectWithoutClassType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use PHPStan\Type\TypeUtils;
+
+final class ArraySearchFunctionDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'array_search';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		if (count($functionCall->args) < 3) {
+			return $functionReflection->getReturnType();
+		}
+
+		$strictArgType = $scope->getType($functionCall->args[2]->value);
+		if (!($strictArgType instanceof ConstantBooleanType) || $strictArgType->getValue() === false) {
+			return $functionReflection->getReturnType();
+		}
+
+		$needleArgType = $scope->getType($functionCall->args[0]->value);
+		$haystackArgType = $scope->getType($functionCall->args[1]->value);
+
+		if (count(TypeUtils::getArrays($haystackArgType)) === 0) {
+			return new NullType();
+		}
+
+		if ($haystackArgType->getIterableValueType()->isSuperTypeOf($needleArgType)->no()) {
+			return new ConstantBooleanType(false);
+		}
+
+		if ($haystackArgType instanceof ConstantArrayType) {
+			return $this->resolveTypeFromConstantHaystackAndNeedle($needleArgType, $haystackArgType);
+		}
+
+		return TypeCombinator::union($haystackArgType->getIterableKeyType(), new ConstantBooleanType(false));
+	}
+
+	private function resolveTypeFromConstantHaystackAndNeedle(Type $needle, ConstantArrayType $haystack): Type
+	{
+		$matchesByType = [];
+
+		/**
+		 * @var int $index
+		 * @var Type $valueType
+		 */
+		foreach ($haystack->getValueTypes() as $index => $valueType) {
+			if ($needle->isSuperTypeOf($valueType)->no()) {
+				continue;
+			}
+
+			if ($needle instanceof ConstantScalarType && $valueType instanceof ConstantScalarType
+				&& $needle->getValue() === $valueType->getValue()
+			) {
+				return $haystack->getKeyTypes()[$index];
+			}
+
+			$matchesByType[] = $haystack->getKeyTypes()[$index];
+		}
+
+		if (count($matchesByType) > 0) {
+			if ($haystack->getIterableValueType()->accepts($needle)
+				&& $needle->isSuperTypeOf(new ObjectWithoutClassType())->no()
```

Comment:
Why is this condition necessary?

Reply:
Because comparison based on object is always risky (can result into false).
I think this condition should be written better, but I don't know how to check if needle is some object, not only without known class name.